From 63e028448bae2a1090e60d798e9cee9cdfbf7156 Mon Sep 17 00:00:00 2001
From: Ryan Zezeski <rpz@joyent.com>
Date: Mon, 4 Feb 2019 14:33:52 -0700
Subject: [PATCH] OS-7556 IPv6 packets dropped after crossing MAC-loopback
 Reviewed by: Robert Mustacchi <rm@joyent.com> Reviewed by: Dan McDonald
 <danmcd@joyent.com> Approved by: Robert Mustacchi <rm@joyent.com>

---
 usr/src/uts/common/inet/ip/ip6_input.c | 15 +++++++++++----
 usr/src/uts/common/inet/ip/ip_input.c  | 21 ++++++++++++++-------
 2 files changed, 25 insertions(+), 11 deletions(-)

diff --git a/usr/src/uts/common/inet/ip/ip6_input.c b/usr/src/uts/common/inet/ip/ip6_input.c
index 79a53eff7c..96cc281da5 100644
--- a/usr/src/uts/common/inet/ip/ip6_input.c
+++ b/usr/src/uts/common/inet/ip/ip6_input.c
@@ -23,7 +23,7 @@
  * Copyright (c) 2009, 2010, Oracle and/or its affiliates. All rights reserved
  *
  * Copyright 2011 Nexenta Systems, Inc. All rights reserved.
- * Copyright 2018 Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 /* Copyright (c) 1990 Mentat Inc. */
 
@@ -1891,6 +1891,16 @@ ip_input_cksum_v6(iaflags_t iraflags, mblk_t *mp, ip6_t *ip6h,
 		return (B_TRUE);
 	}
 
+	hck_flags = DB_CKSUMFLAGS(mp);
+
+	if (hck_flags & HW_LOCAL_MAC) {
+		/*
+		 * The packet is from a same-machine sender in which
+		 * case we assume data integrity.
+		 */
+		return (B_TRUE);
+	}
+
 	/*
 	 * Revert to software checksum calculation if the interface
 	 * isn't capable of checksum offload.
@@ -1907,9 +1917,6 @@ ip_input_cksum_v6(iaflags_t iraflags, mblk_t *mp, ip6_t *ip6h,
 	 * We apply this for all ULP protocols. Does the HW know to
 	 * not set the flags for SCTP and other protocols.
 	 */
-
-	hck_flags = DB_CKSUMFLAGS(mp);
-
 	if (hck_flags & HCK_FULLCKSUM_OK) {
 		/*
 		 * Hardware has already verified the checksum.
diff --git a/usr/src/uts/common/inet/ip/ip_input.c b/usr/src/uts/common/inet/ip/ip_input.c
index 22c1c74391..e2e7dca22c 100644
--- a/usr/src/uts/common/inet/ip/ip_input.c
+++ b/usr/src/uts/common/inet/ip/ip_input.c
@@ -23,7 +23,7 @@
  * Copyright (c) 2009, 2010, Oracle and/or its affiliates. All rights reserved.
  *
  * Copyright 2011 Nexenta Systems, Inc. All rights reserved.
- * Copyright 2018 Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 /* Copyright (c) 1990 Mentat Inc. */
 
@@ -2243,6 +2243,17 @@ ip_input_cksum_v4(iaflags_t iraflags, mblk_t *mp, ipha_t *ipha,
 		/* No ULP checksum to verify. */
 		return (B_TRUE);
 	}
+
+	hck_flags = DB_CKSUMFLAGS(mp);
+
+	if (hck_flags & HW_LOCAL_MAC) {
+		/*
+		 * The packet is from a same-machine sender in which
+		 * case we assume data integrity.
+		 */
+		return (B_TRUE);
+	}
+
 	/*
 	 * Revert to software checksum calculation if the interface
 	 * isn't capable of checksum offload.
@@ -2259,13 +2270,9 @@ ip_input_cksum_v4(iaflags_t iraflags, mblk_t *mp, ipha_t *ipha,
 	 * We apply this for all ULP protocols. Does the HW know to
 	 * not set the flags for SCTP and other protocols.
 	 */
-	hck_flags = DB_CKSUMFLAGS(mp);
-
-	if ((hck_flags & HCK_FULLCKSUM_OK) || (hck_flags & HW_LOCAL_MAC)) {
+	if (hck_flags & HCK_FULLCKSUM_OK) {
 		/*
-		 * Either the hardware already verified the checksum
-		 * or the packet is from a same-machine sender in
-		 * which case we assume data integrity.
+		 * Hardware has already verified the checksum.
 		 */
 		return (B_TRUE);
 	}
-- 
2.21.0

